sample: openamp: add support for stm32h747i_disco board

cameled opened the pull request:
~~**NOTE**: This PR currently are not ready, still have some problem with M4 booting.~~

[EDIT] how to using test this sample on openamp sample
```
# Build M7 and M4
$ west build -p -b stm32h747i_disco_m7 samples/subsys/ipc/openamp

# flash M7
$ west flash

# flash M4 
$ west flash --build-dir build/openamp_remote-prefix/src/openamp_remote-build
```

cameled commented:
@galak @hubertmis Hi, Did you have meet booting problem when working with open-amp and rpmsg_service samples? On stm32h747i_disco board, when run openamp and rpmsg_service sample, M4 are not boot.

stm32h747i_disco have a M7(master) and M4(remote). M4 not support openocd debug, I add led blinky to it's main function, but it's not blink.

I have check the M4 image on flash and zephyr.bin file, it's same.

carlocaione commented:
@cameled just to let you know there is a discussion going on whether we want to deprecate rpmsg_service in favour of ipc_service.

erwango commented:
@cameled Is that specific to those samples ?

cameled commented:
@erwango Yes, I think this issue are not same as #37827 
I test 8KB array on blinky sample base on this PR, M4 blink fine.

cameled commented:
@carlescufi Thanks for you mentioning. open-amp raw interface are not friendly for end user, If the ipc_service have merged, I will try to make stm32h747i_disco support it.

cameled commented:
@erwango I get once time  M4 boot success case with this PR, no matter power-down or reset, M4 will always boot success. After I re-flash the M4 board, boot failed happened again. 

Also, I meet the large-size array issue again with the blinky sample. 

Looks there has some deeper issues that I can't find out.

cameled commented:
@erwango I disabled the CM4 boot config "BCM4" on STM32Progrmmer->option bytes->User Configuration, then all the things return to normal, no open-amp or large size array(#37827) issues exist.

erwango commented:
> @erwango I disabled the CM4 boot config "BCM4" on STM32Progrmmer->option bytes->User Configuration, then all the things return to normal, no open-amp or large size array(#37827) issues exist.

@ABOSTM, can you check this is the expected configuration?
If yes, should we document this, or even implement a check to detect unsupported configurations?

cameled commented:
Another issue, shared memory not work fine. 

If write val1 to sram4(0x38000000) on CM7, 
```
uint32_t *sp = (uint32_t *)0x38000000;
*sp  = val1;
```
then read val2 from sram4(0x38000000)  on CM4.
```
uint32_t *sp = (uint32_t *)0x38000000;
val2 = *sp;
```
val1 and val2 are not equal ;-)

carlocaione commented:
> Another issue, shared memory not work fine.
> 
> If write val1 to sram4(0x38000000) on CM7,
> 
> ```
> uint32_t *sp = (uint32_t *)0x38000000;
> *sp  = val1;
> ```
> 
> then read val2 from sram4(0x38000000) on CM4.
> 
> ```
> uint32_t *sp = (uint32_t *)0x38000000;
> val2 = *sp;
> ```
> 
> val1 and val2 are not equal ;-)

Cache?

cameled commented:
> > Another issue, shared memory not work fine.
> > If write val1 to sram4(0x38000000) on CM7,
> > ```
> > uint32_t *sp = (uint32_t *)0x38000000;
> > *sp  = val1;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then read val2 from sram4(0x38000000) on CM4.
> > ```
> > uint32_t *sp = (uint32_t *)0x38000000;
> > val2 = *sp;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > val1 and val2 are not equal ;-)
> 
> Cache?

Yes, looks like it's a cache issue.
If I add CONFIG_NOCACHE_MEMORY=n to M7 Kconfig, then shared memory work fine.

cameled commented:
@erwango Finally, open-amp and rpmsg_service samples work fine.

There have two issue block this PR.
1. M4 boot issue
    Fix by disable the CM4 boot config "BCM4" on STM32Progrmmer->option bytes->User Configuration,
2. M7 memory cache issue
    Fix by add CONFIG_NOCACHE_MEMORY=y to M7 Kconfig(thanks @carlocaione).

carlocaione commented:
For (2) you want to take a look at https://github.com/OpenAMP/open-amp/pull/293

ABOSTM commented:
> > @erwango I disabled the CM4 boot config "BCM4" on STM32Progrmmer->option bytes->User Configuration, then all the things return to normal, no open-amp or large size array(#37827) issues exist.
> 
> @ABOSTM, can you check this is the expected configuration?
> If yes, should we document this, or even implement a check to detect unsupported configurations?

It is not the expected behavior, it should work with BCM4 enabled (especially as it is the default configuration out of the box), and it used to work when I worked on this dualcore. So it sounds like a regression

cameled commented:
> > > @erwango I disabled the CM4 boot config "BCM4" on STM32Progrmmer->option bytes->User Configuration, then all the things return to normal, no open-amp or large size array(#37827) issues exist.
> > 
> > 
> > @ABOSTM, can you check this is the expected configuration?
> > If yes, should we document this, or even implement a check to detect unsupported configurations?
> 
> It is not the expected behavior, it should work with BCM4 enabled (especially as it is the default configuration out of the box), and it used to work when I worked on this dualcore. So it sounds like a regression

There have some unknown areas, why open-amp not running code(POST_KERNEL and main) can affect the CM4 booting process.

cameled commented:
Deploy of openamp and rpmsg_service on stm32h747i_disco board have some trick. I will try to add it to the README.rst at tomorrow.
```
# Build M7 and M4
$ west build -p -b stm32h747i_disco_m7 samples/subsys/ipc/openamp

# flash M7
$ west flash

# flash M4 
$ west flash --build-dir build/openamp_remote-prefix/src/openamp_remote-build
```

lenghonglin commented:
the pull request is not counting ?  @cameled

lenghonglin commented:
> For (2) you want to take a look at [OpenAMP/open-amp#293](https://github.com/OpenAMP/open-amp/pull/293)

@cameled  but this version not update with zeohyr

cameled commented:
> the pull request is not counting ? @cameled

I think this PR is ready. 
About the memory cache problem, I disabled the memory cache for M7 Core at openamp sample.

lenghonglin commented:
> > the pull request is not counting ? @cameled
> 
> I think this PR is ready. About the memory cache problem, I disabled the memory cache for M7 Core at openamp sample.

i think disable dcache is not the best way, it  a little rough.

cameled commented:
> > > the pull request is not counting ? @cameled
> > 
> > 
> > I think this PR is ready. About the memory cache problem, I disabled the memory cache for M7 Core at openamp sample.
> 
> i think disable dcache is not the best way, it a little rough.
Disable the cache memory for M7 Core can give us a chance to try this feature. 
I will remove it if open-amp module update from upstream.

lenghonglin commented:
> > > > the pull request is not counting ? @cameled
> > > 
> > > 
> > > I think this PR is ready. About the memory cache problem, I disabled the memory cache for M7 Core at openamp sample.
> > 
> > 
> > i think disable dcache is not the best way, it a little rough.
> 
> Disable the cache memory for M7 Core can give us a chance to try this feature. I will remove it if open-amp module update from upstream.

i see [open-amp release](https://github.com/OpenAMP/open-amp/releases) is already released, maybe will update right now.  @carlocaione

carlocaione commented:
> i see [open-amp release](https://github.com/OpenAMP/open-amp/releases) is already released, maybe will update right now. @carlocaione

Opened https://github.com/zephyrproject-rtos/open-amp/pull/11

lenghonglin commented:
> > i see [open-amp release](https://github.com/OpenAMP/open-amp/releases) is already released, maybe will update right now. @carlocaione
> 
> Opened [zephyrproject-rtos/open-amp#11](https://github.com/zephyrproject-rtos/open-amp/pull/11)

How about libmetal?

carlocaione commented:
> How about libmetal?

What about that? That should be already ok.

lenghonglin commented:
> > How about libmetal?
> 
> What about that? That should be already ok.

[zephyr libmetal](https://github.com/zephyrproject-rtos/libmetal/tree/master/libmetal) and [open-amp libmetal](https://github.com/OpenAMP/libmetal) .  open-amp libmetal use new Zephyr cache API.

carlocaione commented:
> > > How about libmetal?
> > 
> > 
> > What about that? That should be already ok.
> 
> [zephyr libmetal](https://github.com/zephyrproject-rtos/libmetal/tree/master/libmetal) and [open-amp libmetal](https://github.com/OpenAMP/libmetal) . open-amp libmetal use new Zephyr cache API.

https://github.com/zephyrproject-rtos/libmetal/pull/12

lenghonglin commented:
> > > > How about libmetal?
> > > 
> > > 
> > > What about that? That should be already ok.
> > 
> > 
> > [zephyr libmetal](https://github.com/zephyrproject-rtos/libmetal/tree/master/libmetal) and [open-amp libmetal](https://github.com/OpenAMP/libmetal) . open-amp libmetal use new Zephyr cache API.
> 
> [zephyrproject-rtos/libmetal#12](https://github.com/zephyrproject-rtos/libmetal/pull/12)

Thanks : : )

lenghonglin commented:
@carlocaione @cameled  Another problem is that the Zephyr doesn't have a cache interface for the ARM framework。

carlocaione commented:
> @carlocaione @cameled Another problem is that the Zephyr doesn't have a cache interface for the ARM framework。

not sure what you mean with ARM framework. See https://github.com/zephyrproject-rtos/zephyr/blob/main/include/cache.h if `!CONFIG_HAS_EXTERNAL_CACHE` you can hook up the arch functions to manage the cache, otherwise if you have an external cache controler you can define `CONFIG_HAS_EXTERNAL_CACHE` and hook up the proper functions.

lenghonglin commented:
> > @carlocaione @cameled Another problem is that the Zephyr doesn't have a cache interface for the ARM framework。
> 
> not sure what you mean with ARM framework. See https://github.com/zephyrproject-rtos/zephyr/blob/main/include/cache.h if `!CONFIG_HAS_EXTERNAL_CACHE` you can hook up the arch functions to manage the cache, otherwise if you have an external cache controler you can define `CONFIG_HAS_EXTERNAL_CACHE` and hook up the proper functions.

ARM framework means Arm Cortex-M33 , Arm Cortex-M7 ,ect. 

Zephyr only support [Arm64](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm64/core/cache.c), [X86](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/x86/core/cache.c), [ARC](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arc/core/cache.c),  not for Arm Cortex-M.

carlocaione commented:
> Zephyr only support [Arm64](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm64/core/cache.c), [X86](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/x86/core/cache.c), [ARC](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arc/core/cache.c), not for Arm Cortex-M.

Right, still not clear what you are asking here. If you need you can add it and push a new PR for that.

lenghonglin commented:
> > Zephyr only support [Arm64](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm64/core/cache.c), [X86](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/x86/core/cache.c), [ARC](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arc/core/cache.c), not for Arm Cortex-M.
> 
> Right, still not clear what you are asking here. If you need you can add it and push a new PR for that.

My Soc is Cortex-M33, If use new version open-amp and libmetal and enable `WITH_DCACHE_VRINGS`, `WITH_DCACHE_BUFFERS`. 

With Zephyr , Call relationship is `VRING_INVALIDATE`--->`metal_cache_invalidate`--->`sys_cache_data_range`--->`cache_data_range`.  

But , if the Architecture is Cortex-M, these is no have `cache_data_range` Implement . 

I just curious why Cortex's Cache is not implemented, which should be a common feature.

lenghonglin commented:
> > Zephyr only support [Arm64](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm64/core/cache.c), [X86](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/x86/core/cache.c), [ARC](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arc/core/cache.c), not for Arm Cortex-M.
> 
> Right, still not clear what you are asking here. If you need you can add it and push a new PR for that.

I will try to push a new PR.

lenghonglin commented:
> @lenghonglin I have removed the CONFIG_NOCACHE_MEMORY option from open-amp sample.

Here's another issue, if u disable CONFIG_NOCACHE_MEMORY , and you will enable `CONFIG_CACHE_MANAGEMENT`， but Zephyr have not implement arm cortex-m cache interface, so you should enable `CONFIG_HAS_EXTERNAL_CACHE` and implement cache interface for stm32h747.

cameled commented:
Thanks for you advice, I will try to add the interface for M7 Core.

cameled commented:
> > @lenghonglin I have removed the CONFIG_NOCACHE_MEMORY option from open-amp sample.
> 
> Here's another issue, if u disable CONFIG_NOCACHE_MEMORY , and you will enable `CONFIG_CACHE_MANAGEMENT`， but Zephyr have not implement arm cortex-m cache interface, so you should enable `CONFIG_HAS_EXTERNAL_CACHE` and implement cache interface for stm32h747.

Thanks for you advice, I will try to add the interface for M7 Core.

lenghonglin commented:
> > > @lenghonglin I have removed the CONFIG_NOCACHE_MEMORY option from open-amp sample.
> > 
> > 
> > Here's another issue, if u disable CONFIG_NOCACHE_MEMORY , and you will enable `CONFIG_CACHE_MANAGEMENT`， but Zephyr have not implement arm cortex-m cache interface, so you should enable `CONFIG_HAS_EXTERNAL_CACHE` and implement cache interface for stm32h747.
> 
> Thanks for you advice, I will try to add the interface for M7 Core.

I see u are work in shenzhen , i am working in shenzhen too , how about leave a WeChat?

cameled commented:
> > > > @lenghonglin I have removed the CONFIG_NOCACHE_MEMORY option from open-amp sample.
> > > 
> > > 
> > > Here's another issue, if u disable CONFIG_NOCACHE_MEMORY , and you will enable `CONFIG_CACHE_MANAGEMENT`， but Zephyr have not implement arm cortex-m cache interface, so you should enable `CONFIG_HAS_EXTERNAL_CACHE` and implement cache interface for stm32h747.
> > 
> > 
> > Thanks for you advice, I will try to add the interface for M7 Core.
> 
> I see u are work in shenzhen , i am working in shenzhen too , how about leave a WeChat?

OK.

erwango commented on samples/subsys/ipc/rpmsg_service/README.rst:
```diff
@@ -36,6 +36,14 @@ Building the application for v2m_musca_b1
+.. zephyr-app-commands::
+   :zephyr-app: samples/subsys/ipc/rpmsg_service
+   :board: stm32h747i_disco_m7
+   :goals: debug
```
I wonder why the 'goals' is debug here (aside from copy/paste).
Besides, shouldn't we document m4 target should be build/flashed as well ?

cameled replied:
openamp and rpmsg_service sample build m7 and m4 target at same time. 
m7 at 
```
build
```
m4 at
```
build/openamp_remote-prefix/src/openamp_remote-build
```

cameled replied:
I will try to adjust stm32h747 board to fully match it.

cameled replied:
'zephyr-app-commands' in README.rst not support add --build-dir option for west flash only. ':build-dir:' will make west build and flash both add a "--build-dir" option.
The nested remote project in openamp sample, makes it hard to use 'zephyr-app-commands' for stm32h747i_disco_m4. 
I think there have two way can fix it. make 'zephyr-app-commands' more flexible; or make a big change for this sample.

erwango replied:
Right, I've came across this as well on [this board](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/b_u585i_iot02a). I ended up using `   .. code-block:: console`.
Don't hesitate to make use of it as that's important for people to easily reproduce

erwango replied:
Having had a try, I confirm that the README should be fixed as the build procedure is unusual.
At least you should mention:
- running the _m7 build command also buid the _m4
- flashing procedures for both cores

Also, it could be interesting to add a note stating that to get the console working on m4 side you need to disable console on m7 and enable on m4. That is not handy but the fastest way to get console on m4 and that's not obvious for most users I guess. (Ideally, it would be nice having usb based console on m4 enabled by default but that's another topic).

cameled replied:
@erwango Thanks!

arnopo commented on subsys/ipc/rpmsg_service/rpmsg_backend.c:
```diff
@@ -131,6 +131,8 @@ static void virtio_notify(struct virtqueue *vq)
 	uint32_t current_core = sse_200_platform_get_cpu_id();
 
 	status = ipm_send(ipm_handle, 0, current_core ? 0 : 1, 0, 1);
+#elif defined(CONFIG_IPM_STM32_HSEM)
```
Could be useful to explain somewhere ( in a comment or in commit message)that the HSEM mailbox is bi-directional and so that the same channel ID can be used in both directions (i.e. for RX and TX virtqueue).

cameled replied:
That's a good idea. HSEM mailbox doesn't support the mailbox channel model, only can notify to each other. no id select and data transfer.

arnopo commented on samples/subsys/ipc/openamp/boards/stm32h747i_disco_m7.overlay:
```diff
@@ -0,0 +1,19 @@
+	chosen {
+		/*
+		 * shared memory reserved for the inter-processor communication
+		 */
```
a reason to use multi-line comment format? 
could be on one line:
`/* shared memory reserved for the inter-processor communication */`

cameled replied:
No other special reason, change to one line.

arnopo commented on samples/subsys/ipc/rpmsg_service/boards/stm32h747i_disco_m7.overlay:
```diff
@@ -0,0 +1,19 @@
+	chosen {
+		/*
+		 * shared memory reserved for the inter-processor communication
+		 */
```
same remark here

cameled replied:
Done.

arnopo reviewed (dismissed):
minor comments else LGTM

erwango commented:
@cameled I've had a try on my board and unfortunately I can't get it working.
Here is what I can get on m7 side:
```
*** Booting Zephyr OS build v2.7.99-1108-g4eec821986ba  ***
Starting application thread!

OpenAMP[master] demo started
```
On M4:
```
*** Booting Zephyr OS build v2.7.99-1108-g4eec821986ba  ***
Starting application thread!

OpenAMP[remote] demo started
```
Trying to debug (m7 side), it seems to be doing right, waiting in `arch_cpu_idle`.
Any advice on what I can do to see what is going on ?

cameled commented:
> @cameled I've had a try on my board and unfortunately I can't get it working. Here is what I can get on m7 side:
> 
> ```
> *** Booting Zephyr OS build v2.7.99-1108-g4eec821986ba  ***
> Starting application thread!
> 
> OpenAMP[master] demo started
> ```
> 
> On M4:
> 
> ```
> *** Booting Zephyr OS build v2.7.99-1108-g4eec821986ba  ***
> Starting application thread!
> 
> OpenAMP[remote] demo started
> ```
> 
> Trying to debug (m7 side), it seems to be doing right, waiting in `arch_cpu_idle`. Any advice on what I can do to see what is going on ?

Quick fix:
add CONFIG_NOCACHE_MEMORY=y to CM7, there have a cache issue on M7 Core.

erwango commented:
> Quick fix:
add CONFIG_NOCACHE_MEMORY=y to CM7, there have a cache issue on M7 Core.

\o/ working fine now! Thanks
Is there a plan to fix this in another PR ? Otherwise,  what about adding this in `samples/subsys/ipc/openamp/boards/stm32h747i_disco_m7.conf`
This would avoid others to face the same issue.

cameled commented:
> > Quick fix:
> > add CONFIG_NOCACHE_MEMORY=y to CM7, there have a cache issue on M7 Core.
> 
> \o/ working fine now! Thanks Is there a plan to fix this in another PR ? Otherwise, what about adding this in `samples/subsys/ipc/openamp/boards/stm32h747i_disco_m7.conf` This would avoid others to face the same issue.

@carlocaione have merge the openamp cache helpers https://github.com/OpenAMP/open-amp/pull/293 to Zephyr. But there have a interface missing issue report by @lenghonglin https://github.com/zephyrproject-rtos/zephyr/pull/38322#issuecomment-957278351.

erwango commented:
> @carlocaione have merged the openamp cache helpers OpenAMP/open-amp#293 to Zephyr. But there have an interface missing issue report by @lenghonglin #38322 (comment).

What about adding CONFIG_NOCACHE_MEMORY with a note specifying this is a temp workaround waiting the issue to be fixed ? Otherwise I'd prefer we avoid to merge current PR if not working as is.

carlocaione reviewed (changes requested):
> What about adding CONFIG_NOCACHE_MEMORY with a note specifying this is a temp workaround waiting the issue to be fixed ? Otherwise I'd prefer we avoid to merge current PR if not working as is.

Agreed

arnopo commented:
> > What about adding CONFIG_NOCACHE_MEMORY with a note specifying this is a temp workaround waiting the issue to be fixed ? Otherwise I'd prefer we avoid to merge current PR if not working as is.
> 
> Agreed

How CONFIG_NOCACHE_MEMORY  is working? 
Look like  an associated no cacheable section can be defined in the linker...
In this case Defining the shared memory in a specific section would make sense, not only has a temporary work around.
As the rpmsg no copy interface is not used, defining the shared memory in a non cacheable area would be probably more efficient.

cameled commented:
> > > What about adding CONFIG_NOCACHE_MEMORY with a note specifying this is a temp workaround waiting the issue to be fixed ? Otherwise I'd prefer we avoid to merge current PR if not working as is.
> > 
> > 
> > Agreed
> 
> How CONFIG_NOCACHE_MEMORY is working? Look like an associated no cacheable section can be defined in the linker... In this case Defining the shared memory in a specific section would make sense, not only has a temporary work around. As the rpmsg no copy interface is not used, defining the shared memory in a non cacheable area would probably be more efficient.

From [ARM Cortex-M7 Reference Manual](https://documentation-service.arm.com/static/5e906b038259fe2368e2a7bb?token=), section "4.1.3 Initializing and enabling the L1 cache", L1 cache are disabled by default.

stm32h7 enable this at soc_m7.c file with CONFIG_NOCACHE_MEMORY check.
```
#ifndef CONFIG_NOCACHE_MEMORY
	if (!(SCB->CCR & SCB_CCR_DC_Msk)) {
		SCB_EnableDCache();
	}
#endif /* CONFIG_NOCACHE_MEMORY */
```

cameled commented:
@arnopo no cacheable section like #30403? I prefer to implement the cache control for the Cortex-M, which will be a more complete solution. Also found libmetal metal_io_region have mem_flags.

erwango commented:
> stm32h7 enable this at soc_m7.c file with CONFIG_NOCACHE_MEMORY check.

This part is indeed uncorrect and I hope https://github.com/zephyrproject-rtos/zephyr/issues/40146 should help providing a better solution.
You can use it for now, but we should keep in mind this is temporary.
Ideally, we should remove usage of  `NOCACHE_MEMORY` which is not proper here and replace with a H7 cache deactivation specific flag.

cameled commented:
Will update at this weekend.

cameled commented:
> Will update at this weekend.

Sorry for the delay. After check the ipc_service sample, I find there have an ipm channel number requirement.
ipm_stm32_hsem mailbox only have one channel for Tx/Rx.

carlocaione commented:
> Sorry for the delay. After check the ipc_service sample, I find there have an ipm channel number requirement.
> ipm_stm32_hsem mailbox only have one channel for Tx/Rx.

That's fine. You can use just one single channel for tx/rx.

Commit:
samples: openamp: add support for stm32h747i_disco board

This add openamp sample run on stm32h747i_disco board.

The CM7 on stm32h747 support memory cache. If the cache enabled on CM7,
CM4 will receive incorrect data. This just disable the memory cache
on CM7, future user may want to use more flexible way to fix it.

Signed-off-by: HaiLong Yang <cameledyang@pm.me>

cameled commented:
Remove rpmsg_service sample support.

Not support ipc_service, it based on mbox api, stm32h747 current not support it.

This just disable the memory cache on CM7, future user may want to use more flexible way to fix it.

cameled commented:
Close this PR.
